File: solutions/10 Regular Expression Matching [Hard]/sol.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int a = s.size(), b = p.size();
        vector<bool> curr(b + 1, false);

        for (int i=0; i<=a; i++){
            bool prev = curr[0];
            curr[0] = !i;

            for (int j=1; j<=b; j++){
                bool t = curr[j];
                if (p[j - 1] == '*'){
                    curr[j] = curr[j - 2] || (i && curr[j] && (p[j - 2] == '.' || p[j - 2] == s[i - 1]));
                }
                else{
                    curr[j] = i && prev && (p[j - 1] == '.' || p[j - 1] == s[i - 1]);
                }
                prev = t;
            }
        }

        return curr[b];
    }
};

// C++
// 0 ms (beats 100.00%)
// 8.24 mb (beats 68.86%)

File: solutions/25 Reverse Nodes in k-Group [Hard]/sol.py
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if k == 1: return head

        def reverse(fi, la, prev):
            nonlocal head
            if prev:
                prev.next = la
            else:
                head = la
            
            prev, curr = fi, fi.next
            fi.next = la.next

            for _ in range(k - 1):
                curr.next, prev, curr = prev, curr, curr.next
        
        ct = 1
        curr = fi = head
        prev = None

        while curr:
            if ct == k:
                reverse(fi, curr, prev)
                prev = fi
                curr, fi = fi.next, fi.next
                ct = 1
            else:
                curr = curr.next
                ct += 1
        
        return head
    
# Python    
# 46 ms (beats 35.12%)
# 17.49 mb (beats 49.73%)